src/mqtt_client.rs

```diff
@@ -480,55 +432,45 @@ where
+                                }
+                            }
                         } else {
                             log::info!("Discarding `List` without `ResponseTopic`");
+                            return;
                         }
-                        // Response sent with listing.
-                        return;
                     } else {
                         Response::error("`List` already in progress")
                     }
                 }
 
                 Command::Get { path } => {
-                    let mut data = [0u8; MESSAGE_SIZE];
-                    match self.settings.get_json(path, &mut data) {
-                        Err(err) => err.into(),
-                        Ok(len) => {
-                            let mut topic = self.prefix.clone();
-
-                            // Note(unwrap): We check that the string will fit during
-                            // construction.
-                            topic
-                                .push_str("/settings")
-                                .and_then(|_| topic.push_str(path))
-                                .unwrap();
-
-                            // Note(unwrap): This construction cannot fail because there's always a
-                            // valid topic.
-                            let message = minimq::Publication::new(&data[..len])
-                                .reply(properties)
-                                // Override the response topic with the path.
-                                .topic(&topic)
-                                .qos(QoS::AtLeastOnce)
-                                .finish()
-                                .unwrap();
-
-                            if client.publish(message).is_err() {
-                                Response::error("Can't publish `Get` response")
-                            } else {
-                                Response::ok()
-                            }
-                        }
+                    let props = [minimq::Property::UserProperty(
+                        minimq::types::Utf8String("code"),
+                        minimq::types::Utf8String(ResponseCode::Ok.as_ref()),
+                    )];
+
+                    let Ok(message) =
+                        DeferredPublication::new(|buf| self.settings.get_json(path, buf))
+                            .reply(properties)
+                            .properties(&props)
+                            // Override the response topic with the path.
+                            .qos(QoS::AtLeastOnce)
+                            .finish()
+                    else {
+                        return;
+                    };
+
+                    match client.publish(message) {
+                        Err(minimq::PubError::Serialization(error)) => error.into(),
+
+                        // Otherwise, we should consider the response sent.
+                        _ => return,
```

Comment:
Could you comment what this returns from? I can't keep the return/continue/break behavior w.r.t match/closures/loops in my head.